### ISSUE TITLE
Fix S907 FP: Don't use `goto case`

### ISSUE THREAD

**Issue by DaleStan** [SonarSource/sonar-dotnet#9763]
### Description

(False positive, I guess? This template fits better than the new rule template.) The existing S907 `goto `_`label`_ rule also triggers on `goto case`/`goto default`.

I have a switch statement with (currently) 26 string cases, akin to the one below. For the sake of finding the labels and reducing merge conflicts when new cases are added, the case labels are in alphabetical order. This requires either code duplication or `goto case` statements. It would be nice to be able to configure `goto case`/`goto default` separately from `goto `_`label`_.

### Reproducer

```c#
switch (insectType) {
    case "ant":
        goto case "termite";
    case "bee":
        goto case "wasp";
    case "termite":
        // handle ants and termites
        break;
    case "wasp":
        // handle bees and wasps
        break;
}
```

In my case, none of the case blocks with `goto case` contain any other code, but it might also be good to have a third setting for cases terminated with goto case, e.g.
```c#
    case "bee":
        // handle bee stuff that doesn't apply to wasps
        goto case "wasp";
    // ...
    case "wasp":
        // handle bees and wasps
        // (everything that applies to wasps also applies to bees)
        break;
```

### Product and Version

SonarQube 8.27.0.14275 and Visual Studio 17.14.16

**Comment by jilles-sg** [comment]
I agree that it makes little sense that case fallthrough is allowed in e.g. C and JavaScript (best add a comment) but not in C# because the spelling for it contains the four letters `goto`. This would, however, only apply to jumping to the immediately following case. Jumping to other cases would be the exact issue that S907 is intended to catch.

In this particular case, I would personally give up the alphabetical order and use `case "ant": case "termite":` (on two lines) or `case "ant" or "termite":`.

**Comment by DaleStan** [comment]
I still believe S907 is too general. `goto `_`label`_ can jump to arbitrary locations in the current method, while `goto case` and `goto default` can only jump to valid `case` labels in the current `switch` statement. I can't jump to a dummy `case null` label when switching on an enum value, nor can I jump to a label in an enclosing switch statement.

**Comment by pavel-mikula-sonarsource** [comment]
Hi,
The `goto case xxx` are intentionally raised by this rule. Those 26 cases of different strings going up and down the list, with possibility of multi-jumps (intentional or not), where some of them are single forwards, some of them are complex handlers, and hopefully non of them is a mix, does not sounds like a code that is easy to read and maintain.

At the same time, we're avoiding parameters as much as we can to avoid parameter hell experience on our rules.

Non-alphabetical multi-case, or alphabetical solution that would just call (or return) a correct handler method for sounds like a cleaner solution here.

I don't know the exact code, so I'll improvise a bit just to illustrate the idea
```
var handler = insectType switch {
    "ant" => woodDestroyingHandler,
    "bee" => yellowStingyHandler,
    "termite" => woodDestroyingHandler,
    "wasp" => yellowStingyHandler,
    _ => throw new UnexpectedInsectException("Who's this anyway?")
}
```

For these reasons, we won't change the behavior of the rule. If this file is an exception in your code base, you could disable the rule with `#pragma warning disable S907` in that file, or around that method.
